packages/plugin-js/src/index.ts

```diff
@@ -103,33 +142,48 @@ export default function pluginJS(options?: Options): Plugin {
         meta: {},
         modes: {},
       };
-      const ts: TSResult = {tokens: [], meta: [], modes: []};
+      const ts: TSResult = {tokens: {}, meta: {}, modes: {}};
       const transform = (typeof options?.transform === 'function' && options.transform) || defaultTransform;
       for (const token of tokens) {
-        js.tokens[token.id] = await transform(token);
+        setToken(js.tokens, token.id, await transform(token), options?.deep);
         if (buildTS) {
           const t = tokenTypes[token.$type];
-          ts.tokens.push(indent(`${objKey(token.id)}: ${t}['$value'];`, 1));
+          setToken(ts.tokens, token.id, `${t}['$value']`, options?.deep);
           tsImports.add(t);
         }
-        js.meta![token.id] = {
-          _original: cloneDeep(token._original),
-          ...((token._group && {_group: token._group}) || {}),
-          ...cloneDeep(token as any),
-        };
+        setToken(
+          js.meta!,
+          token.id,
+          {
+            _original: cloneDeep(token._original),
+            ...((token._group && {_group: token._group}) || {}),
+            ...cloneDeep(token as any),
+          },
+          options?.deep,
+        );
         if (buildTS) {
           const t = `Parsed${tokenTypes[token.$type]}`;
-          ts.meta!.push(indent(`${objKey(token.id)}: ${t}${token.$extensions?.mode ? ` & { $extensions: { mode: typeof modes['${token.id}'] } }` : ''};`, 1));
+          const modeAccessor = options?.deep ? token.id.replace('.', "']['") : token.id;
+          setToken(ts.meta!, token.id, `${t}${token.$extensions?.mode ? ` & { $extensions: { mode: typeof modes['${modeAccessor}'] } }` : ''}`, options?.deep);
           tsImports.add(t);
         }
         if (token.$extensions?.mode) {
-          js.modes[token.id] = {};
-          if (buildTS) ts.modes.push(indent(`${objKey(token.id)}: {`, 1));
+          setToken(js.modes, token.id, {}, options?.deep);
+          if (buildTS) setToken(ts.modes, token.id, {}, options?.deep);
           for (const modeName of Object.keys(token.$extensions.mode)) {
-            js.modes[token.id]![modeName] = await transform(token, modeName);
-            if (buildTS) ts.modes.push(indent(`${objKey(modeName)}: ${tokenTypes[token.$type]}['$value'];`, 2));
+            if (options?.deep) {
+              setToken(js.modes, `${token.id}.${modeName}`, await transform(token, modeName), true);
+            } else {
+              js.modes[token.id]![modeName] = await transform(token, modeName);
+            }
+            if (buildTS) {
+              if (options?.deep) {
+                setToken(ts.modes, `${token.id}.${modeName}`, `${tokenTypes[token.$type]}['$value']`, true);
+              } else {
+                (ts.modes[token.id] as TSResult['modes'])[modeName] = `${tokenTypes[token.$type]}['$value']`;
```

Comment:
😅 I’m so sorry you had to wade through the manual `.d.ts` file generation. I don’t think a lot of this code was written as well as I’d like to be easy to refactor / work on.

But this all looks fantastic 🙂

Reply:
Honestly it wasn’t that bad, it was just very specific! You did most of the work with the serializer function 😄

Reply:
For a lot of codegen stuff, taking a more AST approach is “cleaner” in some ways. But you pay for it in other ways in terms of upfront timesink / wading through the land of obscure bugs, not to mention when you set it down for a while, the confusion of “wait how does this thing work again?”

String mashing is sloppy but it’s real easy to compare “output good” vs “output bad” and know exactly where to make the adjustment